instructions/src/lib.rs

```diff
@@ -1,162 +1,1532 @@
+        match *op {
+            Op::Add => 0x00,
+            Op::Sub => 0x10,
+            Op::Mul => 0x20,
+            Op::Div => 0x30,
+            Op::BitOr => 0x40,
+            Op::BitAnd => 0x50,
+            Op::LShift => 0x60,
+            Op::LogicalRShift => 0x70,
+            Op::Negate => 0x80,
+            Op::Mod => 0x90,
+            Op::BitXor => 0xa0,
+            Op::NoOp => 0xb0,
+            Op::SignRShift => 0xc0
+        }
+    }
+}
+
+pub enum Arch {
+    X64,
+    X32
+}
+
+impl AsRef<Arch> for Arch {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl<'a> From<&'a Arch> for u8 {
+    fn from(arch: &'a Arch) -> u8 {
+        match *arch {
+            Arch::X64 => 0x07,
+            Arch::X32 => 0x04
+        }
+    }
+}
+
+pub struct SwapBytes {
+    endian: Endian,
+    dst: u8,
+    src: u8,
+    offset: u16,
+    imm: u32
+}
+
+impl SwapBytes {
+    pub fn new(endian: Endian) -> Self {
+        SwapBytes {
+            endian: endian,
+            dst: 0x00,
+            src: 0x00,
+            offset: 0x00_00,
+            imm: 0x00_00_00_00
+        }
+    }
+}
+
+impl Instruction for SwapBytes {
+    fn opt_code_byte(&self) -> u8 {
+        self.endian.as_ref().into()
+    }
+
+    fn dst(&self) -> u8 {
+        self.dst
+    }
+
+    fn src(&self) -> u8 {
+        self.src
+    }
+
+    fn offset(&self) -> u16 {
+        self.offset
+    }
+
+    fn imm(&self) -> u32 {
+        self.imm
+    }
+
+    fn dst_reg(mut self, dst: u8) -> Self {
+        self.dst = dst;
+        self
+    }
+
+    fn src_reg(mut self, src: u8) -> Self {
+        self.src = src;
+        self
+    }
+
+    fn offset_bytes(mut self, offset: u16) -> Self {
+        self.offset = offset;
+        self
     }
 
-    pub fn bind_register(&'p mut self) -> Operation<'p> {
-        Operation::new(self, 0xbf)
+    fn immediate(mut self, imm: u32) -> Self {
+        self.imm = imm;
+        self
     }
+}
+
+pub enum Endian {
+    Little,
+    Big
+}
 
-    pub fn load_mem(&'p mut self, size: MemSize) -> Operation<'p> {
-        let opcode = match size {
-            MemSize::Double => 0x79,
-            MemSize::Half => 0x69
-        };
-        Operation::new(self, opcode)
+impl AsRef<Endian> for Endian {
+    fn as_ref(&self) -> &Self {
+        self
     }
+}
 
-    pub fn exit(&'p mut self) -> Operation<'p> {
-        Operation::new(self, 0x95)
+impl<'e> From<&'e Endian> for u8 {
+    fn from(endian: &'e Endian) -> u8 {
+        match *endian {
+            Endian::Little => 0xd4,
+            Endian::Big => 0xdc
+        }
     }
 }
 
-pub struct Operation<'p> {
-    program: &'p mut ProgramCodeBuilder,
-    op: u8,
+pub struct Load {
+    mem_size: MemSize,
+    addressing: Addressing,
+    address_source: AddressSource,
     dst: u8,
     src: u8,
     offset: u16,
-    immediate: u32
+    imm: u32
+}
+
+impl Load {
+    pub fn new(mem_size: MemSize, addressing: Addressing, address_source: AddressSource) -> Self {
```

Comment:
Probably easier to decouple the different kinds of load into several operation type instead of using this `Adressing` everywhere? Maybe also for the `AddressSource` I prefer what you had in your v1?